extensions/sql-database-projects/src/controllers/projectController.ts

```diff
@@ -248,5 +251,212 @@ export class ProjectsController {
+
+			// Add generated file structure to the project
+			await project.addToProject(fileFolderList);
+
+			//Refresh project to show the added files
+			this.refreshProjectsTree();
+		}
+		catch (err) {
+			this.apiWrapper.showErrorMessage(utils.getErrorMessage(err));
+		}
+	}
+
+	private async getProjectName(dbName: string): Promise<string | undefined> {
+		let projName = await this.apiWrapper.showInputBox({
+			prompt: constants.newDatabaseProjectName,
+			value: `DatabaseProject${dbName}`
+		});
+
+		projName = projName?.trim();
+
+		return projName;
+	}
+
+	private async getExtractTarget(): Promise<ExtractTarget> {
+		let extractTarget: ExtractTarget;
+
+		const extractTargetOptions: QuickPickItem[] = [];
+
+		let keys: string[] = Object.keys(ExtractTarget).filter(k => typeof ExtractTarget[k as any] === 'number');
+
+		keys.forEach(targetOption => {
+			let pascalCaseTargetOption: string = utils.toPascalCase(targetOption);	// for better readability
+			extractTargetOptions.push({ label: pascalCaseTargetOption });
+		});
+
+		let input = await this.apiWrapper.showQuickPick(extractTargetOptions.slice(1), {		//Ignore the first option to create Dacpac
+			canPickMany: false,
+			placeHolder: constants.extractTargetInput
+		});
+		let extractTargetInput = input?.label;
+
+		if (extractTargetInput) {
+			let camelCaseInput: string = utils.toCamelCase(extractTargetInput);
+			extractTarget = ExtractTarget[camelCaseInput as keyof typeof ExtractTarget];
+		} else {
+			extractTarget = -1;
+		}
+
+		return extractTarget;
+	}
+
+	private async getFolderLocation(extractTarget: ExtractTarget): Promise<Uri | undefined> {
+		let selectionResult;
+		let projUri;
+
+		if (extractTarget !== ExtractTarget.file) {
+			selectionResult = await this.apiWrapper.showOpenDialog({
+				canSelectFiles: false,
+				canSelectFolders: true,
+				canSelectMany: false,
+				openLabel: constants.selectFileFolder,
+				defaultUri: this.apiWrapper.workspaceFolders() ? (this.apiWrapper.workspaceFolders() as WorkspaceFolder[])[0].uri : undefined
+			});
+			if (selectionResult) {
+				projUri = (selectionResult as Uri[])[0];
+			}
+		} else {
+			// Get filename
+			selectionResult = await this.apiWrapper.showSaveDialog(
+				{
+					defaultUri: this.apiWrapper.workspaceFolders() ? (this.apiWrapper.workspaceFolders() as WorkspaceFolder[])[0].uri : undefined,
+					saveLabel: constants.selectFileFolder,
+					filters: {
+						'SQL files': ['sql'],
+						'All files': ['*']
+					}
+				}
+			);
+			if (selectionResult) {
+				projUri = selectionResult as unknown as Uri;
+			}
+		}
+
+		return projUri;
+	}
+
+	private async isDirEmpty(newProjFolderUri: string): Promise<boolean> {
+		return (await fs.readdir(newProjFolderUri)).length === 0;
+	}
+
+	private async importApiCall(model: ImportDataModel): Promise<void> {
+		let ext = extensions.getExtension(mssql.extension.name);
+
+		if (ext === undefined) {
+			this.apiWrapper.showErrorMessage('VSCode extension undefined');
```

Comment:
i would say better to check before using.. no?

Reply:
1. The error message refers to VSCode extension - not sure what that's supposed to mean
2. If the msssql extension isn't there there's something else horribly wrong so it probably isn't worth adding extra code to check it since we're in a bad state anyways.

Reply:
getExtension can return an undefined result as well, hence added an extra check.
And yes, the error message is indeed incorrect- that should be updated.

Reply:
It *can* but if it ever does we're in serious trouble. The check is never going to do anything useful since if it isn't there for some reason nothing else will work regardless.

Reply:
I strongly believe we should check before using. I am surprised the code will even let you build without warning if a variable is not checked before using.

Reply:
I'm with Udeesha here.  If "but if it ever does, we're in serious trouble" is true, then the accessor for extensions should throw instead of possibly returning `undefined`.

I think a `!` assertion here gives the best of both worlds without changing the `getExtension()` method signature:
`let ext: ExtensionClassName = extensions.getExtension(mssql.extension.name)!;`

Reply:
Yeah, the code doesn't like not been checked for the undefined variable. And that is precisely why this check was added at the first place.

Reply:
Well, the accessor shouldn't throw because in the normal case an extension might not exist so it's not reasonable to throw. It's in this specific case where the MSSQL extension itself not existing because it's built in (and specifically a hugely core part of our product) that I'm saying this is unnecessary.

What Benjin said with the ! assertion would be the way around the error from the strict null checking. 

Don't get me wrong - I'm all for checking undefined if it's necessary. But if this ever returned undefined then nothing is going to work anyways so I just don't see the point in adding extra code if we can't do anything about it.

(I liken this case to a System Exception like OutOfMemory error in a C# app. Sure you could catch it and then do...what?)